File: ruby/lib/natsuzora.rb
# frozen_string_literal: true

require_relative 'natsuzora/version'
require_relative 'natsuzora/errors'

module Natsuzora
  class << self
    def render(source, data, include_root: nil)
      require_ruby!
      Template.new(source, include_root: include_root).render(data)
    end

    def parse(source, include_root: nil)
      require_ruby!
      Template.new(source, include_root: include_root)
    end

    private

    def require_ruby!
      return if defined?(@ruby_loaded)

      require_relative 'natsuzora/token'
      require_relative 'natsuzora/validator'
      require_relative 'natsuzora/html_escape'
      require_relative 'natsuzora/value'
      require_relative 'natsuzora/ast'
      require_relative 'natsuzora/lexer'
      require_relative 'natsuzora/parser'
      require_relative 'natsuzora/context'
      require_relative 'natsuzora/template_loader'
      require_relative 'natsuzora/renderer'
      @ruby_loaded = true
    end
  end
end

require_relative 'natsuzora/template'

File: typescript/src/natsuzora.ts
/**
 * Natsuzora - A Minimal Template Language
 *
 * Public API for parsing and rendering templates.
 */

import { Template } from "./ast.ts";
import { tokenize } from "./lexer.ts";
import { parse as parseTokens } from "./parser.ts";
import { render as renderAst, TemplateLoader } from "./renderer.ts";
import { createTemplateLoader } from "./template_loader.ts";
import { readFileSync } from "./platform.ts";

export interface RenderOptions {
  /**
   * Root directory for include paths.
   * Required when using includes.
   */
  includeRoot?: string;
}

export interface ParsedTemplate {
  ast: Template;
  templateLoader: TemplateLoader | null;
}

/**
 * Parse a template source string into an AST.
 */
export function parse(source: string): Template {
  const tokens = tokenize(source);
  return parseTokens(tokens);
}

/**
 * Parse a template with include support.
 */
export function parseWithIncludes(source: string, includeRoot: string): ParsedTemplate {
  const ast = parse(source);
  const templateLoader = createTemplateLoader(includeRoot);
  return { ast, templateLoader };
}

/**
 * Render a parsed template with the given data.
 */
export function renderTemplate(
  template: ParsedTemplate,
  data: Record<string, unknown>
): string {
  return renderAst(template.ast, data, template.templateLoader);
}

/**
 * Render a template source string with the given data.
 *
 * This is the main entry point for rendering templates.
 */
export function render(
  source: string,
  data: Record<string, unknown>,
  options: RenderOptions = {}
): string {
  const ast = parse(source);

  let templateLoader: TemplateLoader | null = null;
  if (options.includeRoot) {
    templateLoader = createTemplateLoader(options.includeRoot);
  }

  return renderAst(ast, data, templateLoader);
}

/**
 * Render a template file with the given data.
 */
export function renderFile(
  filePath: string,
  data: Record<string, unknown>,
  options: RenderOptions = {}
): string {
  const source = readFileSync(filePath);
  return render(source, data, options);
}

// Re-export types and utilities
export type { Template } from "./ast.ts";
export type { TemplateLoader } from "./renderer.ts";
export { createTemplateLoader } from "./template_loader.ts";

// Re-export error types
export {
  NatsuzoraError,
  LexerError,
  ParseError,
  ReservedWordError,
  RenderError,
  UndefinedVariableError,
  TypeError,
  IncludeError,
  ShadowingError,
} from "./errors.ts";
